src/Engine/ProtoCore/DSASM/HeapObjects.cs

```diff
@@ -18,11 +24,28 @@ public DSArray(int size, Heap heap)
             MetaData = new MetaData { type = (int)PrimitiveType.kTypeArray };
         }
 
-        public IEnumerable<StackValue> Values
+        /// <summary>
+        /// Create an array and populuate with input values
+        /// </summary>
+        /// <param name="values"></param>
+        /// <param name="heap"></param>
+        public DSArray(StackValue[] values, Heap heap)
+            : base(values, heap)
+        {
+            Dict = new Dictionary<StackValue, StackValue>();
```

Comment:
Yes. It is hybrid - integer based key will directly use array, otherwise switch to dictionary. Moving forward, depend on whether we allow to have holes in the array or not (i.e., `x = {}; x[1000000] = 1;` so what is `x[0]`?), we could even put some integer key to dictionary.